File: python/problems/leetcode/0122_best_time_to_buy_and_sell_stock_ii/solution.py
"""LeetCode 122: Best Time to Buy and Sell Stock II."""

from __future__ import annotations


class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        profit = 0
        for i in range(1, len(prices)):
            if prices[i] > prices[i - 1]:
                profit += prices[i] - prices[i - 1]
        return profit


if __name__ == "__main__":
    assert Solution().maxProfit([7, 1, 5, 3, 6, 4]) == 7
    assert Solution().maxProfit([1, 2, 3, 4, 5]) == 4
    assert Solution().maxProfit([7, 6, 4, 3, 1]) == 0
    print("0122 OK")

File: cpp/problems/leetcode/0122_best_time_to_buy_and_sell_stock_ii/solution.cpp
// LeetCode 122
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit = 0;
        for (int i = 1; i < (int)prices.size(); ++i)
            if (prices[i] > prices[i - 1])
                profit += prices[i] - prices[i - 1];
        return profit;
    }
};

int main() {
    vector<int> a{7, 1, 5, 3, 6, 4};
    assert(Solution().maxProfit(a) == 7);
    vector<int> b{1, 2, 3, 4, 5};
    assert(Solution().maxProfit(b) == 4);
    vector<int> c{7, 6, 4, 3, 1};
    assert(Solution().maxProfit(c) == 0);
    cout << "0122 OK" << endl;
    return 0;
}
